bluemira/equilibria/coils/_field.py

```diff
@@ -0,0 +1,770 @@
+            semianalytic_Bz, x, z, split, coil_x, coil_z, coil_dx, coil_dz
+        )
+
+    def F(self, eqcoil: CoilGroup) -> np.ndarray:  # noqa :N802
+        """
+        Calculate the force response at the coil centre including the coil
+        self-force.
+
+        .... math::
+
+             \\mathbf{F} = \\mathbf{j}\\times \\mathbf{B}
+            F_x = IB_z+\\dfrac{\\mu_0I^2}{4\\pi X}\\textrm{ln}\\bigg(\\dfrac{8X}{r_c}-1+\\xi/2\\bigg)
+            F_z = -IBx
+        """  # noqa :W505
+        multiplier = self.current * 2 * np.pi * self.x
+        cr = self._current_radius
+        if any(cr != 0):
+            # true divide errors for zero current coils
+            cr_ind = np.where(cr != 0)
+            fx = np.zeros_like(cr)
+            fx[cr_ind] = (
+                MU_0
+                * self.current[cr_ind] ** 2
+                / (4 * np.pi * self.x[cr_ind])
+                * (np.log(8 * self.x[cr_ind] / cr[cr_ind]) - 1 + 0.25)
+            )
+        else:
+            fx = 0
+
+        return np.array(
+            [
+                multiplier * (eqcoil.Bz(self.x, self.z) + fx),
+                -multiplier * eqcoil.Bx(self.x, self.z),
+            ]
+        ).T
+
+    def control_F(self, coil: CoilGroup) -> np.ndarray:  # noqa :N802
+        """
+        Returns the Green's matrix element for the coil mutual force.
+
+        \t:math:`Fz_{i,j}=-2\\pi X_i\\mathcal{G}(X_j,Z_j,X_i,Z_i)`
+        """
+        # TODO Vectorise
+        x, z = np.atleast_1d(self.x), np.atleast_1d(self.z)  # single coil
+        pos = np.array([x, z])
+        response = np.zeros((x.size, coil.x.size, 2))
+        coils = coil._coils
+        for j, coil2 in enumerate(coils):
+            xw = np.where(x == coil2.x)[0]
+            zw = np.where(z == coil2.z)[0]
+            same_pos = np.where(xw == zw)[0]
+            if same_pos.size > 0:
+                # self inductance
+                xxw = xw[same_pos]
+                cr = self._current_radius[xxw]
+                Bz = np.zeros((x.size, 1))
+                Bx = Bz.copy()  # Should be 0 anyway
+                mask = np.zeros_like(Bz, dtype=bool)
+                mask[same_pos] = True
+                if any(cr != 0):
+                    cr_ind = np.where(cr != 0)
+                    Bz[mask][cr_ind] = (
+                        MU_0
+                        / (4 * np.pi * x[cr_ind])
+                        * (np.log(8 * x[cr_ind] / cr[cr_ind]) - 1 + 0.25)
+                    )
+                if False in mask:
+                    Bz[~mask] = coil2.unit_Bz(*pos[:, ~mask[:, 0]])
+                    Bx[~mask] = coil2.unit_Bx(*pos[:, ~mask[:, 0]])
+
+            else:
+                Bz = coil2.unit_Bz(x, z)
+                Bx = coil2.unit_Bx(x, z)
+            # 1 cross B
+            response[:, j, :] = (
+                2 * np.pi * x[:, None] * np.squeeze(np.array([Bz, -Bx]).T)
+            )
+        return response
+
+
+class CoilFieldsMixin(CoilGroupFieldsMixin):
```

Comment:
Lot of duplication here, is this needed? Only get rid of it if the increased coupling is not too bad..

Reply:
This was the minimal I could do while storing the coordinates of a single coils as floats otherwise we will sometimes have arrays and sometimes floats and its hard to track. The first two functions are unfortunate dupes but the third is just a wrapper.